File: VSCode_Java/leetcode/editor/en/397.integer-replacement.java
//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public int integerReplacement(int n) {
        int ans = 0;
        while (n > 1){
            if((n & 1) == 0){
                n >>= 1;
                ans += 1;
            } else if (n == 3) {
                ans += 2;
                n = 1;
            } else if ((n & 3) == 3) {
                n = n / 2 +1;
                ans += 2;
            } else {
                n = n/2 ;
                ans += 2;
            }
        }
        return ans;
    }
}

//leetcode submit region end(Prohibit modification and deletion)

File: 22FALL/397.integer-replacement.py
#
# @lc app=leetcode id=397 lang=python3
#
# [397] Integer Replacement
#

# @lc code=start

class Solution:
    def integerReplacement(self, n: int) -> int:
        ans = 0
        while n!=1:
            if n % 2 == 0:
                ans += 1
                n //= 2
            elif n % 4 == 1:
                ans += 2
                n //= 2
            else:
                if n == 3:
                    ans += 2
                    n = 1
                else:
                    ans += 2
                    n = n // 2 + 1
        return ans
# @lc code=end

